jest.config.js

```diff
@@ -1,10 +1,33 @@
-module.exports = {
+// For now:
+// - server tests go in the test/ dir and are run in the node environment
+// - client tests go in __tests__ files anywhere else, and are run in the jsdom environment
+//
+// This may not be ideal long-term, but we need a simple pattern-matching way to distinguish
+// between client and server tests. -@jasoncrawford 2019-12-03
+
+const common = {
     preset: "ts-jest",
-    testEnvironment: "jsdom",
     moduleNameMapper: {
         "^(admin|site|charts|utils|db|settings)/(.*)$": "<rootDir>/$1/$2",
         "^settings$": "<rootDir>/settings",
         "^serverSettings$": "<rootDir>/serverSettings"
-    },
-    setupFilesAfterEnv: ["<rootDir>/test/enzymeSetup.ts"]
+    }
+}
+
+module.exports = {
+    projects: [
+        {
+            ...common,
+            displayName: "server",
+            testEnvironment: "node",
+            testMatch: ["<rootDir>/test/**/?(*.)+(spec|test).[jt]s?(x)"]
+        },
+        {
+            ...common,
+            displayName: "client",
+            testEnvironment: "jsdom",
+            setupFilesAfterEnv: ["<rootDir>/test/enzymeSetup.ts"],
+            testMatch: ["**/__tests__/**/*.[jt]s?(x)"]
```

Comment:
I can see this works fine but I don't understand the regex: it's a `j` or `t` followed by an optional `s` and a required `x`? Is it not supposed to be `[jt]sx?` instead?

Reply:
Yeah it's a weird regex. I was confused too. It actually does mean what you think it ought to be saying. Instead of `x?` they use `?(x)`. See https://github.com/micromatch/micromatch#extended-globbing